tempesta_fw/ss_skb.c

```diff
@@ -976,3 +976,105 @@ ss_skb_split(struct sk_buff *skb, int len)
 
 	return buff;
 }
+
+static inline int
+__coalesce_frag(SsSkbList *skb_list, skb_frag_t *frag)
+{
+	struct sk_buff *skb = ss_skb_peek_tail(skb_list);
+
+	if (!skb || skb_shinfo(skb)->nr_frags == MAX_SKB_FRAGS) {
+		skb = ss_skb_alloc();
+		if (!skb)
+			return -ENOMEM;
+		ss_skb_queue_tail(skb_list, skb);
+	}
+
+	skb_shinfo(skb)->frags[skb_shinfo(skb)->nr_frags++] = *frag;
+	ss_skb_adjust_data_len(skb, frag->size);
+	__skb_frag_ref(frag);
+
+	return 0;
+}
+
+static int
+ss_skb_queue_coalesce_tail(SsSkbList *skb_list, const struct sk_buff *skb)
+{
+	int i;
+	skb_frag_t head_frag;
+	unsigned int headlen = skb_headlen(skb);
+
+	if (headlen) {
+		BUG_ON(!skb->head_frag);
+		head_frag.size = headlen;
+		head_frag.page.p = virt_to_head_page(skb->head);
+		head_frag.page_offset = skb->data -
+			(unsigned char *)page_address(head_frag.page.p);
+		if (__coalesce_frag(skb_list, &head_frag))
+			return -ENOMEM;
+	}
+
+	for (i = 0; i < skb_shinfo(skb)->nr_frags; i++) {
+		if (__coalesce_frag(skb_list, &skb_shinfo(skb)->frags[i]))
+			return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int
+ss_skb_unroll_slow(SsSkbList *skb_list, struct sk_buff *skb)
+{
+	struct sk_buff *f_skb;
+
+	if (ss_skb_queue_coalesce_tail(skb_list, skb))
+		goto cleanup;
+
+	skb_walk_frags(skb, f_skb) {
+		if (ss_skb_queue_coalesce_tail(skb_list, f_skb))
+			goto cleanup;
+	}
+
+	consume_skb(skb);
```

Comment:
If `skb_list` is empty, then `__coalesce_frag()` allocates new `skb` and here you free it. It'd be better to optimize the case and consume skb only when we don't need to allocate a new one.